### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by taleldayekh**
Code for the first iteration of the Raiden Installer. Since a lot was stubs in the previous code I decided to start a new. The installer is currently pretty basic and only works with the Görli testnet. It runs in the browser where the user is asked to provide a password and a Infura Project ID before a keystore and config file is created and Raiden is downloaded, extracted and initialized. The installer can also be run as a single file executable.

### CODE REVIEW

**Comment by lullis** [comment]
Your PR is basically overwriting the whole project. This would be okay for most of the source files (as there were only stubs in the original one) but it will pay off to actually keep all of the boilerplate project files (.gitignore, linters, etc) from the original code.

To avoid a series of commits with very tiny corrections and fixes, please take a look at how to "squash commits". [This article](https://medium.com/@vanshbadkul/squashing-commits-in-git-a-step-by-step-guide-884581224adf) might help. Feel free to ask as well.

**Comment by taleldayekh** [comment]
Thank you for taking the time reviewing and for the feedback @lullis. I’ll read the article you linked to after lunch. :)

**Comment by nlsdfnbch** [comment]
Please do not squash commits. The whole point of a VCS is to be able to atomically undo changes in case things go south. Squashing commits makes this impossible.

File: raiden_installer/installer_parts/funding.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+            abi=custom_token_abi
+        )
+        
+    def mint_tokens(
+        self,
+        token_amount: int,
+        gas: int,
+        gas_price: int
+    ) -> tuple:
+        mint = self.custom_token_contract.functions.mint(
```
Comment (nlsdfnbch) [inline raiden_installer/installer_parts/funding.py]: No need to concatenate all these functions to a single line - it does not cost company money to spread code out across several lines and variables.

```suggestion
        mint_result = self.custom_token_contract.functions.mint(token_amount)
        mint_result.buildTransaction(
            {
                'chainId': self.chain_id,
                'gas': gas,
                'gasPrice': gas_price,
                'nonce': self.w3.eth.getTransactionCount(self.address)
            }
        )
```

File: raiden_installer/installer_parts/funding.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        return transaction_details
+
+    def approve_deposit(
+        self,
+        token_amount: int,
+        gas: int,
+        gas_price: int
+    ) -> tuple:
+        approve = self.custom_token_contract.functions.approve(
+            self.user_deposit_address,
```
Comment (nlsdfnbch) [inline raiden_installer/installer_parts/funding.py]: see above.

File: raiden_installer/installer_parts/funding.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        return transaction_details
+
+    def make_deposit(
+        self,
+        token_amount: int,
+        gas: int,
+        gas_price: int
+    ) -> tuple:
+        deposit = self.user_deposit_contract.functions.deposit(
+            self.address,
```
Comment (nlsdfnbch) [inline raiden_installer/installer_parts/funding.py]: see above

File: raiden_installer/installer_parts/keystore.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    Creates a keystore directory with a
+    passphrase encrypted keystore file.
+    '''
+    keystore = Path(keystore_dir).joinpath('keystore')
+    keystore.mkdir(exist_ok=True)
+
+    keyfile = Path(keystore).joinpath(keyfile_name)
+    keyfile_content = create_keyfile_json(os.urandom(32), keystore_pwd.encode())
+
+    with open(keyfile, 'w') as f:
```
Comment (nlsdfnbch) [inline raiden_installer/installer_parts/keystore.py]: Should be:

```suggestion
with keyfile.open("w") as f:
    ...
```

File: raiden_installer/installer_parts/raiden.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
+import os
```
Comment (nlsdfnbch) [inline raiden_installer/installer_parts/raiden.py]: You can rewrite these to either use `contextlib.contextmanager` or the magic methods `__enter__` and `__exit__` of a class , or even a class which will make the code alot cleaner. Take a look at the original code, there's an example of that there.

File: raiden_installer/installer_parts/raiden_config.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
+import os
+import re
+import toml
+from pathlib import Path
+from eth_utils import to_checksum_address
+
+
+class PlainTxtPwd:
```
Comment (nlsdfnbch) [inline raiden_installer/installer_parts/raiden_config.py]: Letters do not cost money, please don't use abbreviations like these - they make it harder to read code, and in the case of more exotic shortenings, promote misunderstandings.

File: raiden_installer/installer_server.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+        static_folder=static_folder
+    )
+else:
+    # Installer code running live
+    app = Flask(__name__)
+
+
+@app.route('/', methods=['GET', 'POST'])
+def install_raiden():
+    if request.method == 'POST':
```
Comment (nlsdfnbch) [inline raiden_installer/installer_server.py]: This function is morbidly obese. Please slim it down - you're already using block comments to separate the code, you might as well outsource them to functions.

File: raiden_installer/constants.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
+import sys
+from pathlib import Path
+from xdg import XDG_DATA_HOME, XDG_CONFIG_HOME
+
+
+'''
```
Comment (nlsdfnbch) [inline raiden_installer/constants.py]: This would fit better as a module docstring, which should be at the top of the file.

File: raiden_installer/constants.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+CONFIG_DIR = Path(XDG_CONFIG_HOME).joinpath('Raiden')
+CONFIG_DIR.mkdir(parents=True, exist_ok=True)
+KEYSTORE_DIR = Path(XDG_DATA_HOME).joinpath('Raiden')
+KEYSTORE_DIR.mkdir(parents=True, exist_ok=True)
+
+
+PLATFORM = 'macOS' if sys.platform == 'darwin' else 'linux'
+
+
+'''
```
Comment (nlsdfnbch) [inline raiden_installer/constants.py]: if you need to section your file using block quotes, consider splitting it up into submodules instead.

File: setup.py [hunk 10]
```diff
@@ -1,60 +1,48 @@
...
-    entry_points={
+    packages=find_packages(exclude=['tests']),
+    install_requires=list_requirements('requirements.txt'),
+    entry_point={
         'console_scripts': [
-            'install-raiden=raiden_installer:main',
-        ],
-    },
-
-    project_urls={
```
Comment (nlsdfnbch) [inline setup.py]: I'd keep the urls for pypi.

**Comment by taleldayekh** [comment]
Thanks @nlsdfnbch appreciate the feedback. I'll work through the pieces bit by bit, haha.

**Comment by taleldayekh** [comment]
Will also read up on how I can add black and isort to my editor.

File: raiden_installer/installer_server.py (hunk 7 above)
Comment (taleldayekh) [reply raiden_installer/installer_server.py]: Argree, this one is intended to be refactored though, just haven't decided in what way yet.

**Comment by lullis** [comment]
> Please do not squash commits. The whole point of a VCS is to be able to atomically undo changes in case things go south. Squashing commits makes this impossible.

@nlsdfnbch, I agree on the larger point; it is better to have commits that are small and easy to revert. My idea was mostly to reduce the number of commits that are just fix for typos or series of commits that should be done as an amend.

**Comment by taleldayekh** [comment]
Thanks again @lullis, I'll go through your input and comment where I feel confused.

File: raiden_installer/constants.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
+import sys
+from pathlib import Path
+from xdg import XDG_DATA_HOME, XDG_CONFIG_HOME
```
[the remark this answers is not in the document]
Comment (taleldayekh) [reply raiden_installer/constants.py]: I debated this with Tore last week. I am generally in favor of reducing the amount dependencies used in a project and was initially about to do what you are suggesting.

However, Tore's input was that if everyone would write their own snippets for such tasks we would end up with a lot of code just repeating some functionality that can be easily attained through some library. Hence, dependencies are not necessarily a bad thing.

File: raiden_installer/constants.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+'''
+Constants for interacting with smart contracts used
+for funding the Pathfinding and Monitoring services.
+'''
+TOKEN_AMOUNT = 10 ** 18
+GAS_PRICE = 2_000_000_000
+
+GAS_REQUIRED_FOR_MINT = 70_000
+GAS_REQUIRED_FOR_APPROVE = 70_000
+GAS_REQUIRED_FOR_DEPOSIT = 200_000
```
[the remark this answers is not in the document]
Comment (taleldayekh) [reply raiden_installer/constants.py]: Did not know that. :) Should I leave one empty line in all files or just constants.py?

File: raiden_installer/constants.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+necessary for initializing Raiden.
+'''
+BINARY_DIR = Path('/usr/local/bin')
+CONFIG_DIR = Path(XDG_CONFIG_HOME).joinpath('Raiden')
+CONFIG_DIR.mkdir(parents=True, exist_ok=True)
+KEYSTORE_DIR = Path(XDG_DATA_HOME).joinpath('Raiden')
+KEYSTORE_DIR.mkdir(parents=True, exist_ok=True)
+
+
+PLATFORM = 'macOS' if sys.platform == 'darwin' else 'linux'
```
[the remark this answers is not in the document]
Comment (taleldayekh) [reply raiden_installer/constants.py]: The PLATFORM variable stores either 'macOS' or 'linux' depending on what sys.platform returns. I need this string to building the Raiden download URL for the correct system. How would I do it otherwise?

We are not supporting Windows? Or am I misunderstanding your comment?

File: raiden_installer/installer_parts/funding.py [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+        f'{mac_address}-{current_user}'.encode()
+    ).hexdigest()
+
+    try:
+        res = requests.post(
+            'https://faucet.workshop.raiden.network/',
+            json={'address': address, 'client_hash': client_hash}
+        )
+        print(res)
+    except requests.exceptions.RequestException as err:
```
[the remark this answers is not in the document]
Comment (taleldayekh) [reply raiden_installer/installer_parts/funding.py]: Your right, I should probably stop the program completely if this is the case. At the moment the program gets stuck in a while loop that keeps querying how much ETH is on the user's account.

File: raiden_installer/installer_parts/raiden.py [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+from zipfile import ZipFile
+from tarfile import TarFile
+
+
+def latest_raiden_release() -> str:
+    try:
+        res = requests.get(
+            'https://api.github.com/repos/raiden-network/raiden/releases'
+        )
+    except requests.exceptions.RequestException as err:
```
[the remark this answers is not in the document]
Comment (taleldayekh) [reply raiden_installer/installer_parts/raiden.py]: This is just a temporary solution, I’ve opened an issue to handle these as logs, but need to read up on it a bit first.

File: raiden_installer/installer_parts/funding.py [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+        print(err)
+
+
+class PfsAndMonitoringFunding:
```
[the remark this answers is not in the document]
Comment (taleldayekh) [reply raiden_installer/installer_parts/funding.py]: This would be a good question to discuss. I made it a class mostly to organize the code but also not to create tons of functions that return something that then needs to be passed somewhere else etc. But I’m not sure this is the best approach and would love to hear feedback on a better solution.

File: raiden_installer/installer_server.py [hunk 17]
```diff
@@ -0,0 +1,220 @@
+import sys
+import json
+import time
+import socket
+import webbrowser
+from pathlib import Path
+from web3 import Web3, HTTPProvider
```
[the remark this answers is not in the document]
Comment (taleldayekh) [reply raiden_installer/installer_server.py]: Agree, received the same feedback from Nils. I’m intending to refactor this but have not decided on what way to go about doing it yet. Basically, everything was dumped there for now to see if all pieces worked together as intended.

File: raiden_installer/installer_server.py [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+    # Installer code running live
+    app = Flask(__name__)
+
+
+@app.route('/', methods=['GET', 'POST'])
+def install_raiden():
+    if request.method == 'POST':
+        keystore_pwd = request.form['keystore-pwd']
+        proj_id = request.form['proj-id']
+        network = 'goerli'
```
[the remark this answers is not in the document]
Comment (taleldayekh) [reply raiden_installer/installer_server.py]: Do you mean the user input generated from the html form? Or just the variables, `keystore_pwd = None`, like so?

File: raiden_installer/installer_server.py [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+
+        '''
+        Installer Step 1
```
[the remark this answers is not in the document]
Comment (taleldayekh) [reply raiden_installer/installer_server.py]: Agree, the comments in this case are also just a part of temporarily putting things together and see the app running.

File: raiden_installer/installer_server.py [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+        '''
+        Installer Step 5
+
+        Create a plain txt pwd file
+        and generate a TOML config.
+        '''
+        plain_txt_pwd = raiden_config.PlainTxtPwd(CONFIG_DIR, keystore_pwd)
+        plain_txt_pwd.create_plain_txt_pwd_file()
+
+        global config_file
```
[the remark this answers is not in the document]
Comment (taleldayekh) [reply raiden_installer/installer_server.py]: I posted in the dev-questions channel on Rocket.Chat about this and decided for this approach as a way to pass data across requests.

**Comment by taleldayekh** [comment]
> Please take a look at the comments below and let me know how I can help you. On the large, the most fundamental changes I'd like to see are related to having a better separation between business-logic (as in "what needs to happen in the program") and presentation ("how the user interacts with the program").
> 
> The tests are also somewhat confusing. The main focus of a test suite is making the developers more confident that the application works as intended and designed. What you've done surely have helped you to figure out how to write tests, but it seems that the things being tested are not really relevant and don't exercise the use cases that might be problematic.
> 
> Just as an example, it will be very important to have tests that can answer things like the following:
> 
> * What happens if the user aborts the install process in the middle of the run and tries to run things again?
> * What happens if the user runs the installer after it has already set up once? Will people lose their already-minted tokens?
> * How easy can we change the test network?
> * What happens if the user provides invalid information?
> * etc etc

Great feedback!
I might need to sit down with you though to get a better understanding of how to do the separation. :)

Also, those are relevant questions and I’ve been thinking about them as well. But the objectives for this first iteration was to get something up and running and not bother creating a context manager that remembers user settings or the installation process.

As you know, we agreed on only supporting Görli for now. But if you are thinking about how easy it is to change testnet in the code it should just be a matter of changing the hardcoded value for ”network” in installer_server.py. I initially made a radio button for this but got feedback on leaving it out for now.

I've written some JS that handles invalid user input in the frontend and there is Python code preventing anything else than a hexadecimal string as a Project ID. If the user provides an invalid Project ID that happens to be a hexadecimal nothing much is done at the moment, unfortunately. :)

Basically, if the user aborts the install process or starts the installer a second time everything gets created a new. This is not a problem at the moment because as I've understood it the purpose of the installer is to easily try Raiden, making transactions, etc. Of course, depending on the roadmap for the installer these thinks might need to be taken into consideration in the future.

A bigger problem right now though is that the user needs to wait for 1h after a failed installation before acquiring ETH again. I haven't had time to think about what would be the appropriate thing to do here, maybe you have a suggestion?

File: raiden_installer/constants.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (lullis) [reply raiden_installer/constants.py]: The issue I am pointing out here is that you are defining a variable that is (a) being used in only one place (the function that builds the URL), (b) whose name (`PLATFORM`) indicates something that already something that python interpreter can give you and (c) slightly differs from the standard only for the sake of the function that you are implementing.

All in all, the overall point is that this does not need to be a "constant" in the program. It will be better to remove this definition altogether and just have the function that builds to URL to contain its own logic about how to build the URL.

[the remark this answers is not in the document]
Comment (lullis) [reply raiden_installer/constants.py]: Re: "not supporting windows", you are right that we are not supporting it at the moment.

However, it is always good to be wary of possible changes and the impact on the code you are creating. By having your code rely on the condition that "it is never going to run on windows", you are making the application very fragile to changes.

File: raiden_installer/installer_server.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (lullis) [reply raiden_installer/installer_server.py]: I meant the `network = 'goerli'` line. If Görli is the only network we currently support, it would be better to have this indicated in the `constants` module than in the middle of a function that should not be responsible in defining anything.

File: raiden_installer/installer_server.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (lullis) [reply raiden_installer/installer_server.py]: But you don't need to pass data across requests, do you? It seems you just want to be able to read/write to the same file. Unless I misunderstood the functionality, the only thing that needs to be known is the name of the file, so that you can open/close it freely.

File: raiden_installer/installer_parts/keystore.py (hunk 4 above)
Comment (taleldayekh) [reply raiden_installer/installer_parts/keystore.py]: Curious to know the difference because in the docs I only seem to find the `open(file, mode)` or is this way specific for just ETH keyfiles?

Comment (taleldayekh) [reply raiden_installer/installer_parts/keystore.py]: Just got clarified that it belongs to the Path object, so never mind my question. :)
